// dodopayments/src/resources/customers/wallets/wallets.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from "../../../resource.js";
import * as Core from "../../../core.js";
import * as MiscAPI from "../../misc.js";
import * as LedgerEntriesAPI from "./ledger-entries.js";
import {
  CustomerWalletTransaction,
  CustomerWalletTransactionsDefaultPageNumberPagination,
  LedgerEntries,
  LedgerEntryCreateParams,
  LedgerEntryListParams,
} from "./ledger-entries.js";

export class Wallets extends APIResource {
  ledgerEntries: LedgerEntriesAPI.LedgerEntries = new LedgerEntriesAPI.LedgerEntries(this._client);

  list(customerId: string, options?: Core.RequestOptions): Core.APIPromise<WalletListResponse> {
    return this._client.get(`/customers/${customerId}/wallets`, options);
  }
}

export interface CustomerWallet {
  balance: number;

  created_at: string;

  currency: MiscAPI.Currency;

  customer_id: string;

  updated_at: string;
}

export interface WalletListResponse {
  items: Array<CustomerWallet>;

  /**
   * Sum of all wallet balances converted to USD (in smallest unit)
   */
  total_balance_usd: number;
}

Wallets.LedgerEntries = LedgerEntries;
Wallets.CustomerWalletTransactionsDefaultPageNumberPagination =
  CustomerWalletTransactionsDefaultPageNumberPagination;

export declare namespace Wallets {
  export { type CustomerWallet as CustomerWallet, type WalletListResponse as WalletListResponse };

  export {
    LedgerEntries as LedgerEntries,
    type CustomerWalletTransaction as CustomerWalletTransaction,
    CustomerWalletTransactionsDefaultPageNumberPagination as CustomerWalletTransactionsDefaultPageNumberPagination,
    type LedgerEntryCreateParams as LedgerEntryCreateParams,
    type LedgerEntryListParams as LedgerEntryListParams,
  };
}
